packages/core/src/utils/cursor.ts

```diff
@@ -0,0 +1,146 @@
+import {
+  type AnyColumn,
+  type AnyColumnWithTable,
+  type OperandValueExpression,
+  type SelectQueryBuilder,
+  sql,
+} from "kysely"
+import { z } from "zod"
+
+export interface PaginatedResult<T> {
+  data: T[]
+  next: Cursor | null
+}
+
+// Cursors are opaque strings that are parsed in the individual repositories
+export const CursorSchema = z.string()
+
+export const PaginateInputSchema = z
+  .object({
+    take: z.number(),
+    cursor: CursorSchema.optional().nullable(),
+  })
+  .optional()
+  .default({ take: 20, cursor: undefined })
+
+export type Pageable = z.infer<typeof PaginateInputSchema>
+export type Cursor = z.infer<typeof CursorSchema>
+
+type GetNextCursorOptions<T> = {
+  pageable: Pageable
+  buildCursor: (values: unknown[]) => Cursor
+}
+
+type PaginatedQueryOptions<DB, TB extends keyof DB, C extends AnyColumn<DB, TB> | AnyColumnWithTable<DB, TB>, O> = {
+  pageable: Pageable
+  order: "asc" | "desc"
+  decodeCursorOverride?: (cursor: Cursor) => OperandValueExpression<DB, TB, C>[]
+  buildCursorOverride?: (record: O) => Cursor
+  columns: C[]
+}
+
+/**
+ * Add cursor based pagination to a query
+ *
+ * Cursors are opaque strings that represent the next record that will be fetched in the ordered set of records based on the given column.
+ * A cursor value of `null` will return no records, and is returned when there are no extra records to fetch.
+ * A cursor value of `undefined` will return the first `take` records.
+ *
+ * For more complex queries that require ordering by multiple columns, manual implementation
+ * is needed.
+ */
+export async function paginatedQuery<
+  DB,
+  TB extends keyof DB,
+  O,
+  C extends AnyColumn<DB, TB> | AnyColumnWithTable<DB, TB>,
+>(query: SelectQueryBuilder<DB, TB, O>, options: PaginatedQueryOptions<DB, TB, C, O>) {
+  // Validate args
+  const {
+    pageable: { cursor, take },
+    order,
+  } = options
+
+  if (cursor === null) {
+    return {
+      next: null,
+      data: [],
+    }
+  }
+
+  const decodeCursor = options.decodeCursorOverride ?? defaultDecodeCursor
+  // Take N+1 to determine if there is a record behind `take`.
+  let pagedQuery = query.limit(take + 1)
+
+  const columnSql = sql`(${sql.raw(options.columns.map((col) => `"${col}"`).join(","))})`
+  pagedQuery = pagedQuery.orderBy(columnSql, order)
+
+  if (cursor !== undefined) {
+    // biome-ignore lint/style/noNonNullAssertion: see id default vals logic above
+    const decodedCursor = decodeCursor!(cursor)
+    // pagedQuery = pagedQuery.where(options.column[0], order === "asc" ? ">=" : "<=", decodedCursor[0])
+    const _columns = sql`(${sql.raw(options.columns.map((col) => `"${col}"`).join(","))})`
+    const _values = sql`(${sql.join(decodedCursor)})`
```

Comment:
```suggestion
    const _values = sql.join(decodedCursor)
```
Does this not work? (this also needs some cleanup)

Reply:
Nope, that does not work. See https://old.kyse.link/?p=s&i=4OIEEWLXFhzj4T3b4n0I

Reply:
You can use an expression builder (but you unfortunately cannot pass a rest to either eb.tuple or eb.tupleRef due to the types

```ts
await db
  .selectFrom('person')
  .select(['id', 'first_name'])
  .where(({ eb }) => eb(eb.refTuple(columns[0], columns[1]), '>=', eb.tuple(sql.val(values[0]), sql.val(values[1]))))
  .execute()
```

```sql
SELECT
  "id",
  "first_name"
FROM
  "person"
WHERE
  ("created_at", "id") >= ($1, $2)

-- Parameters
-- [1] 123123
-- [2] 123123
```